crates/pallet-domains/src/staking.rs

```diff
@@ -387,119 +559,365 @@ pub(crate) fn do_deregister_operator<T: Config>(
-                let domain_stake_summary =
-                    DomainStakingSummary::<T>::get(operator.current_domain_id)
-                        .ok_or(Error::DomainNotInitialized)?;
+                *maybe_withdrawals = Some(withdrawals);
+                Ok(())
+            })
+        })
+    })
+}
 
-                let total_stake = match domain_stake_summary.current_epoch_rewards.get(&operator_id)
-                {
-                    None => operator.current_total_stake,
-                    Some(rewards) => {
-                        let operator_tax = operator.nomination_tax.mul_floor(*rewards);
-                        operator
-                            .current_total_stake
-                            .checked_add(rewards)
-                            .ok_or(Error::BalanceOverflow)?
-                            // deduct operator tax
-                            .checked_sub(&operator_tax)
-                            .ok_or(Error::BalanceUnderflow)?
-                    }
-                };
+/// Unlocks any withdraws that are ready to be unlocked.
+pub(crate) fn do_unlock_funds<T: Config>(
+    operator_id: OperatorId,
+    nominator_id: NominatorId<T>,
+) -> Result<BalanceOf<T>, Error> {
+    Withdrawals::<T>::try_mutate_exists(operator_id, nominator_id.clone(), |maybe_withdrawals| {
+        let withdrawals = maybe_withdrawals.as_mut().ok_or(Error::MissingWithdrawal)?;
+        let Withdrawal {
+            allowed_since_domain_epoch,
+            shares,
+        } = withdrawals.pop_front().ok_or(Error::MissingWithdrawal)?;
+
+        // if there are no withdrawals, then delete the storage as well
+        if withdrawals.is_empty() {
+            *maybe_withdrawals = None
+        }
 
-                let nominator_share =
-                    Perbill::from_rational(nominator.shares, operator.total_shares);
+        let (domain_id, epoch_idx) = allowed_since_domain_epoch.deconstruct();
 
-                let nominator_staked_amount = nominator_share.mul_floor(total_stake);
+        let epoch_completed_block =
+            DomainEpochCompleteAt::<T>::get(domain_id, epoch_idx).ok_or(Error::EpochNotComplete)?;
 
-                let nominator_remaining_amount = nominator_staked_amount
-                    .checked_sub(&withdraw_amount)
-                    .ok_or(Error::BalanceUnderflow)?;
+        let unlock_block_number = epoch_completed_block
+            .checked_add(&T::StakeWithdrawalLockingPeriod::get())
+            .ok_or(Error::BlockNumberOverflow)?;
 
-                if operator_owner == nominator_id {
-                    // for operator owner, the remaining amount should not be less than MinimumOperatorStake,
-                    if nominator_remaining_amount < T::MinOperatorStake::get() {
-                        return Err(Error::MinimumOperatorStake);
-                    }
+        // this is the oldest unconfirmed domain block
+        // we want to release funds only after the unlock period is confirmed.
```

Comment:
Why? This way the effective time until the withdrawal gets to the user is until epoch end+locking period+challenge period now?

Reply:
Ahh yes, this need updated as well

Reply:
Should be updated now